### ISSUE TITLE
Support Workspace Trust

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode-jupyter#5745]
Filed here https://github.com/microsoft/vscode-python/issues/16126
I'm filing this in case they forget our extension.

Hello :wave: I'm from the VS Code team.

Recently, we have been exploring a security feature we refer to as _Workspace Trust_. This feature is intended to centralize and unify a security conscious decision required by a variety of VS Code features. With workspace trust, the user will be able to declare whether or not they trust the folder that is opened in VS Code before these features are executed.

### Why you should care

Your extension is _incredibly popular_ with VS Code users! We want to make sure that those users have a delightful experience with workspace trust and that includes extension authors deciding how much of their extension is supported in an untrusted workspace.

### Workspace Trust experience

You can enable the feature with the following setting `security.workspace.trust.enabled`. Once enabled, you will see the following dialog when opening folders in VS Code.

![Workspace Trust Startup Dialog](https://media.githubusercontent.com/media/microsoft/vscode-docs/vnext/release-notes/images/1_56/trust-dialog.png)

This dialog is important for allowing the user to make a decision early and understand the impact of their decision. Once you understand the feature, you may want to customize when to display the dialog using the setting `security.workspace.trust.startupPrompt`.

You can follow the development of Workspace Trust and provide feedback in [issue #106488](https://github.com/microsoft/vscode/issues/106488).

### Workspace trust API

First off, all of what I am about to say can be found in [issue #120251](https://github.com/microsoft/vscode/issues/120251). That issue will include discussion of the feature and any updates to the feature.

The Workspace Trust extension API is now in stable. This allowed us to release the first cut of our guide for onboarding your extension to Workspace Trust. The API is small, so here is a quick look.

You can declare your extension to provide complete, partial or no support in untrusted workspaces using the `untrustedWorkspaces` capability in `package.json`.

The following example declares that the extension is supported completely in untrusted workspaces. In this case, the extension is enabled in untrusted workspaces.

```json
"capabilities": {
  "untrustedWorkspaces": {
    "supported": true
  }
}
```

The next example declares that the extension is not supported in untrusted workspaces. In this case, the extension is disabled in untrusted workspaces.

```json
"capabilities": {
  "untrustedWorkspaces": {
    "supported": false
  }
}
```

The third option is to declared `limited` support. There are three tools provided to you when you select the `limited` option.

First, if you have a setting that can be configured in the workspace but requires the workspace to be trusted in order to apply the workspace value, then you can include the setting using `restrictedConfigurations` array property in `untrustedWorkspaces` object. Doing so, VS Code will ignore the workspace value of these restricted settings when your extension reads these settings values using the VS Code [Workspace Configuration](https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration) API.

The following example declares the settings that are restricted in untrusted workspaces.

```json
"capabilities": {
  "untrustedWorkspaces": {
    "supported": "limited",
    "restrictedConfigurations": [
      "markdown.styles"
    ]
  }
}
```

Next, you can also check and listen if the current workspace is trusted or not programmatically using the following API:

```ts
export namespace workspace {
  /**
   * When true, the user has explicitly trusted the contents of the workspace.
   */
  export const isTrusted: boolean;
  /**
   * Event that fires when the current workspace has been trusted.
   */
  export const onDidGrantWorkspaceTrust: Event<void>;
}
```

Lastly, you can hide commands or views declaratively with the `isWorkspaceTrusted` context key in your when clauses.

A far more detailed guide on how to onboard which will be updated as we receive feedback can be found in [issue #120251](https://github.com/microsoft/vscode/issues/120251).

### Rollout plan

Workspace Trust will remain disabled for the month of May, but we are planning on enabling this by default in the future. To prepare for that day, we would love for you to try it out and provide feedback.

### We'd love your feedback

Since this issue was created in an automated fashion, we won't be monitoring the responses in this issue (our notifications would explode!). Instead we ask you to drop questions, and feedback in [issue #120251](https://github.com/microsoft/vscode/issues/120251) as we've mentioned above.

We're excited to see what you do with workspace trust!

**Comment by DonJayamanne** [comment]
The API/workflow could change.
We'll need to discuss with VS Code before working on this.

**Comment by DonJayamanne** [comment]
@rebornix Please see comments here https://github.com/microsoft/vscode-jupyter/issues/5938#issuecomment-847119136

**Comment by rchiodo** [comment]
I think there's still customer development work to be done even when the workspace is trusted. Do users understand that opening a notebook could execute code?

**Comment by DonJayamanne** [comment]
> Do users understand that opening a notebook could execute code?

Not sure what you mean.

By execution of code, I'm assuming you mean `execution of rendered output`

**Comment by rchiodo** [comment]
Yes the javascript in the rendered output (as well as markdown, although I think we should just scrape that even in trusted notebooks like collab? does)

Do customers even think that a notebook can do something nasty just by opening it.

**Comment by rchiodo** [comment]
I mean maybe there's a 2nd option for notebooks. Trust but don't let any javascript run or something like that.

**Comment by DonJayamanne** [comment]
If workspace is not trusted:
* We support partial trusted workspaces (to allow opening & viewing notebooks)
* Interactive window
	* Hide run cell (debug) & similar code lenses that run code
	* Ensure interactive window cannot be opened when not trusted
* WebView notebooks
	* Hide the trusted icon
	* Disable changing kernels
	* Hide run & related buttons in webview notebooks
	* Do we render the output by using old trust functionality (i.e. if worksapce is not trusetd, assume notebook is not trusted and use old code to achieve laoding of notebooks & hiding output) - easier than refactoring everything.
	* When a workspace is trusted, trigger trusting of the notebook model (easier than refactoring everything)
* Native Notebooks
	* Hide trust icon (remove old implementation of trust for native notebooks)
	* No need to do anything special in Native Notebooks
* Miscellaneous
	* Disable opening interactive window (in case we miss something, e.g. we miss a command)
	* Hide run commands in command palette (Use appropriate when clauses)
	* Disable auto start
	* Native notebooks - Disable kernel fetching (return empty list).
	* Once trusted, trigger fetching kernels for native notebooks.
* Python Extension API
	* They will not load when workspace is not trusted.
	* If user trusts a worksapce, then extension gets activated later (hence some of our code might have to change, e.g. fetching kernels, etc).
* When workspace is trusted, re-enable the stuff that was disabled
	* Basically we should ensure users don't have to re-load window
	* Re-enable code lenses
	* Re-enable commands (this will happen automatically via when clauses, hence no changes to support this)
	* Re-enable run buttons in webviews

**Comment by rchiodo** [comment]
This still doesn't address the question of whether a user will expect javascript code to run just on open. Even when trusted. I suspect most people don't realize this fact. They don't worry at all about opening notebooks because, hey I'm just opening it.

Trusted workspaces is basically making this even more obtuse because the decision is on a whole workspace. They might have not even realized they had trusted the workspace 3 weeks ago and now they open a notebook.

Maybe you're referring to this?

> We support partial trusted workspaces (to allow opening & viewing notebooks)

What's a partially trusted workspace? What does Jupyter do for 'untrusted' notebooks? Do they scrape javascript from outputs until you run them?

**Comment by DonJayamanne** [comment]
> They might have not even realized they had trusted the workspace 3 weeks ago and now they open a notebook.

Same applies to other extensions. 
E.g. i might have trusted a worksapce 3 weeks ago, and now I just open a workspace and open the test explorer for  python, that ends up running code (automatically).
Or i could trust a worksapce 3 weeks ago and do a `git pull` on some repo that ends up pulling some malicious python code..

**Comment by DonJayamanne** [comment]
Either way, this is a generic problem that needs to be addressed by the core VS Code product.

**Comment by rchiodo** [comment]
> Either way, this is a generic problem that needs to be addressed by the core VS Code product.

Or we could handle it for notebooks at least. I think it wouldn't be horrible if notebooks didn't run javascript until you ran a cell.

**Comment by DonJayamanne** [comment]
> we could handle it for notebooks at least. I think it wouldn't be horrible if notebooks didn't run javascript until 

I believe this is what VS Code products hopes to avoid, have a unified experience for all. If something doesnt work, then tweak their product to handle such cases.

**Comment by DavidKutu** [comment]
Can't run cells, open interactive window, commands and code lenses are hidden when the workspace is untrusted.
Validated
